google-cloud-firestore/src/test/java/com/google/cloud/firestore/LocalFirestoreHelper.java

```diff
@@ -929,22 +929,45 @@ public static String autoId() {
     return fromJsonString(json.replace("'", "\""));
   }
 
-  static class RequestResponseMap
+  /** Map of request/response pairs to be returned when `sendRequest()` is called. */
+  static class RequestResponseStubber
       extends LinkedHashMap<GeneratedMessageV3, ApiFuture<? extends GeneratedMessageV3>> {
+    void initializeStub(
+        ArgumentCaptor<? extends Message> argumentCaptor, FirestoreImpl firestoreMock) {
+      Stubber stubber = null;
+      for (final ApiFuture<? extends GeneratedMessageV3> response : values()) {
+        stubber = (stubber != null) ? stubber.doReturn(response) : doReturn(response);
+      }
+      if (stubber != null) {
+        stubber
+            .when(firestoreMock)
+            .sendRequest(argumentCaptor.capture(), Matchers.<UnaryCallable<Message, Message>>any());
+      }
+    }
+  }
+
+  /**
+   * Map of request/response pairs to be returned when `sendRequest()` is called.
+   *
+   * <p>Enforces that only one active request can be pending at a time.
+   */
+  static class SerialRequestStubber extends RequestResponseStubber {
     int activeRequestCounter = 0;
     SettableApiFuture<Void> activeRequestComplete = SettableApiFuture.create();
     CountDownLatch latch = new CountDownLatch(1);
 
-    void markComplete() {
+    void markRequestComplete() {
       activeRequestComplete.set(null);
     }
 
-    void awaitLatch() {
+    void awaitRequest() {
+      activeRequestComplete = SettableApiFuture.create();
       try {
         latch.await();
       } catch (Exception e) {
         fail("latch.await() should not fail");
       }
+      latch = new CountDownLatch(1);
```

Comment:
I think you have to use semaphores here to allow for multiple requests to build up and be consumed. The state needs to span the lifetime of the Stubber.

Reply:
Done.